bigml/predicate_utils/utils.py

```diff
@@ -0,0 +1,184 @@
+        full_term = full_term.lower()
+
+    return 1 if text == full_term else 0
+
+
+def get_tokens_flags(case_sensitive):
+    """Returns flags for regular expression matching depending on text analysis
+    options
+    """
+    flags = re.U
+    if not case_sensitive:
+        flags = (re.I | flags)
+
+    return flags
+
+
+def term_matches_tokens(text, forms_list, case_sensitive):
+    """Counts the number of occurences of the words in forms_list in the text
+    """
+    flags = get_tokens_flags(case_sensitive)
+    expression = r'(\b|_)%s(\b|_)' % '(\\b|_)|(\\b|_)'.join(forms_list)
+    pattern = re.compile(expression, flags=flags)
+    matches = re.findall(pattern, text)
+    return len(matches)
+
+
+def item_matches(text, item, options):
+    """ Counts the number of occurences of the item in the text
+    The matching considers the separator or
+    the separating regular expression.
+    """
+    separator = options.get('separator', ' ')
+    regexp = options.get('separator_regexp')
+    if regexp is None:
+        regexp = r"%s" % re.escape(separator)
+
+    return count_items_matches(text, item, regexp)
+
+
+def count_items_matches(text, item, regexp):
+    """ Counts the number of occurences of the item in the text
+    """
+    expression = r'(^|%s)%s($|%s)' % (regexp, item, regexp)
+    pattern = re.compile(expression, flags=re.U)
+    matches = re.findall(pattern, text)
+
+    return len(matches)
+
+def apply_predicates(node, input_data, fields):
+    num_predicates = node[1]
+
+    for i in range(num_predicates):
+        operation = node[2 + (5 * i)]
+        field = node[3 + (5 * i)]
+        value = node[4 + (5 * i)]
+        term = node[5 + (5 * i)]
+        missing = node[6 + (5 * i)]
+
+        if not apply_predicate(operation, field, value, term, missing,
+                               input_data, fields[field]):
+            return False
+
+    return True
+
+
+def apply_predicate(operator, field, value, term, missing, input_data,
+                    field_info):
+    """Applies the operators defined in the predicate as strings to
+    the provided input data
+    """
+    # for missing operators
+    if input_data.get(field) is None:
+        # text and item fields will treat missing values by following the
+        # doesn't contain branch
+        if term is None:
+            return missing or (
+                operator == EQ and value is None)
+    elif operator == NE and value is None:
+        return True
+
+    if term is not None:
+        if field_info['optype'] == 'text':
+            all_forms = field_info['summary'].get('term_forms', {})
+            term_forms = all_forms.get(term, [])
+            terms = [term]
+            terms.extend(term_forms)
+            options = field_info['term_analysis']
+            return OPERATOR[operator](*[term_matches(
+                input_data.get(field, ""),
+                terms, options),
+                                             value])
```

Comment:
a super nit, but  find this very difficult to read... since you already have a few local vars defined, how about using one for `term_matches(...)`? i think it'd be clearer...